### DISCUSSION TITLE
push %rbx being the first instruction can lead to confusion

### DISCUSSION THREAD

**Discussion by youzhongyang** [openzfs/zfs#11728] [Q&A] [answered]
The following assembly code functions have the push %rbx as the first instruction:

sha1_block_data_order
SHA256TransformBlocks
SHA512TransformBlocks

https://github.com/openzfs/zfs/blob/master/module/icp/asm-x86_64/sha1/sha1-x86_64.S#L72
https://github.com/openzfs/zfs/blob/master/module/icp/asm-x86_64/sha2/sha256_impl.S#L86
https://github.com/openzfs/zfs/blob/master/module/icp/asm-x86_64/sha2/sha512_impl.S#L87

Normally it is expected to have the %rbp pushed first, if instead %rbx is pushed first, some stack traversing code may get confused.

I got these messages on one of our servers:
```
[Wed Mar 10 15:38:53 2021] WARNING: kernel stack regs at 0000000015d4b0a1 in z_null_iss:2002 has bad 'bp' value 00000000895aaaf9
[Wed Mar 10 15:38:53 2021] unwind stack type:0 next_sp:0000000001d965b6 mask:0x6 graph_idx:0
```
Any comment?

**Comment by AttilaFueloep** [comment, marked as answer]
This code uses %rbp as a general purpose register, not as the frame pointer. It behaves like gcc with the -fomit-frame-pointer option. So pushing %rbp first won't help since later on %rbp won't point to the previous %rbp. One could add appropriate CFI directives since it seems that the needed DWARF sections are generated for regular (non-debug) build as well. I can have a stab at that if it's deemed important.

**Reply by behlendorf** [reply]
@AttilaFueloep I think this would definitely be worthwhile to address.  It also causes a warning from `objtool` when building, see issue #6950.

**Reply by AttilaFueloep** [reply]
All right, will open a PR soon. It won't fix the `objtool` warnings though. I'll comment over there.

**Reply by AttilaFueloep** [reply]
#11733

**Reply by AttilaFueloep** [reply]
I think this could be marked as resolved/answered.
